File: ocaml/lib/code_analysis/evaluator.ml
open Runtime.Value
open Binding.Node.Expression

let rec evaluate_expression vars node =
  match node with
  | Assignment a ->
      let value = evaluate_expression vars a.assignment_expression in
      Hashtbl.add vars a.assignment_variable value;
      value
  | Binary b -> (
      let left = evaluate_expression vars b.binary_left in
      let right = evaluate_expression vars b.binary_right in
      match b.binary_op.bop_kind with
      | Addition -> Int (get_int left + get_int right)
      | Subtraction -> Int (get_int left - get_int right)
      | Multiplication -> Int (get_int left * get_int right)
      | Division -> Int (get_int left / get_int right)
      | LogicalAnd -> Bool (get_bool left && get_bool right)
      | LogicalOr -> Bool (get_bool left || get_bool right)
      | Equality -> Bool (left = right)
      | Inequality -> Bool (left = right |> not))
  | Literal l -> l.literal_value
  | Unary u -> (
      let operand = evaluate_expression vars u.unary_operand in
      match u.unary_op.uop_kind with
      | Identity -> operand
      | Negation -> Int (-get_int operand)
      | LogicalNegation -> Bool (get_bool operand |> not))
  | Variable v -> Hashtbl.find vars v.variable_variable

let evaluate variables root = evaluate_expression variables root

File: nim/src/minskpkg/codeAnalysis/evaluator.nim
import minskpkg/minskObject

import binding/expressions/[
  boundAssignmentExpression,
  boundBinaryExpression,
  boundBinaryOperator,
  boundLiteralExpression,
  boundUnaryExpression,
  boundUnaryOperator,
  boundVariableExpression,
]
import binding/[
  boundExpression,
  boundNode,
  boundNodeKind,
]
import variableMap

type
  Evaluator* = object
    root: BoundExpression
    variables: VariableMap

proc newEvaluator*(
  root: BoundExpression,
  variables: VariableMap
): Evaluator =
  result.root = root
  result.variables = variables

proc evaluateExpression(evaluator: var Evaluator, root: BoundExpression): MinskObject =
  case root.kind
  of BoundNodeKind.AssignmentExpression:
    let a = root.BoundAssignmentExpression
    let value = evaluator.evaluateExpression(a.expression)
    evaluator.variables[a.variable] = value
    return value
  of BoundNodeKind.LiteralExpression:
    let l = root.BoundLiteralExpression
    return l.value
  of BoundNodeKind.BinaryExpression:
    let b = root.BoundBinaryExpression
    let left = evaluateExpression(evaluator, b.left)
    let right = evaluateExpression(evaluator, b.right)
    case b.op.kind
    of BoundBinaryOperatorKind.Addition:
      return moInteger(left.intVal + right.intVal)
    of BoundBinaryOperatorKind.Subtraction:
      return moInteger(left.intVal - right.intVal)
    of BoundBinaryOperatorKind.Multiplication:
      return moInteger(left.intVal * right.intVal)
    of BoundBinaryOperatorKind.Division:
      return moInteger(left.intVal div right.intVal)
    of BoundBinaryOperatorKind.LogicalAnd:
      return moBoolean(left.boolVal and right.boolVal)
    of BoundBinaryOperatorKind.LogicalOr:
      return moBoolean(left.boolVal or right.boolVal)
    of BoundBinaryOperatorKind.Equality:
      return moBoolean(left == right)
    of BoundBinaryOperatorKind.Inequality:
      return moBoolean(left != right)
  of BoundNodeKind.UnaryExpression:
    let u = root.BoundUnaryExpression
    let operand = evaluateExpression(evaluator, u.operand)
    case u.op.kind
    of BoundUnaryOperatorKind.Identity:
      return moInteger(operand.intVal)
    of BoundUnaryOperatorKind.Negation:
      return moInteger(-operand.intVal)
    of BoundUnaryOperatorKind.LogicalNegation:
      return moBoolean(not operand.boolVal)
  of BoundNodeKind.VariableExpression:
    let v = root.BoundVariableExpression
    return evaluator.variables[v.variable]

proc evaluate*(evaluator: var Evaluator): MinskObject =
  evaluator.evaluateExpression(evaluator.root)
